compiler/expressions.go

```diff
@@ -1130,12 +1136,17 @@ func (c *funcContext) translateImplicitConversion(expr ast.Expr, desiredType typ
 			// wrap JS object into js.Object struct when converting to interface
 			return c.formatExpr("new $jsObjectPtr(%e)", expr)
 		}
+
+		switch exprType.Underlying().(type) {
```

Comment:
As I understand it, we effectively need some combination of the logic from `translateImplicitConversionWithCloning` and `translateImplicitConversion` when it comes to array/struct values.

As this PR stands:

* `translateImplicitConversionWithCloning` handles the case where the target is an array/struct type which by definition means the source must be exactly the same type, hence a `$clone` is required
* `translateImplicitConversion` handles the conversion from an array/struct type to an interface type, hence the `new X` is required to wrap the `$clone`-ed value (part of #661 concerned the fact that we weren't cloning before wrapping)

I'm relatively relaxed on where the code lives because once your first comment is addressed I think this PR looks simpler as far as `translateImplicitConversion` is concerned in any case.

Let me know what you think